File: CS/ValidateCell_MVVM/MainViewModel.cs
using DevExpress.Mvvm;
using DevExpress.Mvvm.DataAnnotations;
using DevExpress.Mvvm.Xpf;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;

namespace ValidateCell_MVVM {
    public class Product {
        public string ProductName { get; set; }
        public double UnitPrice { get; set; }
        public bool Discontinued { get; set; }

        public Product(int i) {
            ProductName = "Product" + i;
            Discontinued = i % 5 == 0;
        }
    }
    public class MainViewModel : ViewModelBase {
        public ObservableCollection<Product> ProductList { get; }

        public MainViewModel() {
            ProductList = new ObservableCollection<Product>(GetProductList());
        }
        static IEnumerable<Product> GetProductList() {
            Random r = new Random();
            return Enumerable.Range(0, 20).Select(i => new Product(i) { UnitPrice = r.Next(1, 50) });
        }

        [Command]
        public void ValidateCell(RowCellValidationArgs args) {
            if(args.FieldName != nameof(Product.UnitPrice) || !((Product)args.Item).Discontinued)
                return;
            var cellValue = (double)args.OldValue;
            var discount = 100 - Convert.ToDouble(args.NewValue) / cellValue * 100;
            if(discount > 0 && discount <= 30)
                return;

            var error = discount < 0
                ? $"The price cannot be greater than {cellValue}"
                : $"The discount cannot be greater than 30% ({cellValue * 0.7}). Please correct the price.";
            args.Result = new ValidationErrorInfo(error, ValidationErrorType.Critical);
        }
    }
}

File: VB/ValidateCell_MVVM/MainViewModel.vb
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.DataAnnotations
Imports DevExpress.Mvvm.Xpf
Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.Linq

Namespace ValidateCell_MVVM

    Public Class Product

        Public Property ProductName As String

        Public Property UnitPrice As Double

        Public Property Discontinued As Boolean

        Public Sub New(ByVal i As Integer)
            ProductName = "Product" & i
            Discontinued = i Mod 5 = 0
        End Sub
    End Class

    Public Class MainViewModel
        Inherits ViewModelBase

        Public ReadOnly Property ProductList As ObservableCollection(Of Product)

        Public Sub New()
            ProductList = New ObservableCollection(Of Product)(GetProductList())
        End Sub

        Private Shared Function GetProductList() As IEnumerable(Of Product)
            Dim r As Random = New Random()
            Return Enumerable.Range(0, 20).[Select](Function(i) New Product(i) With {.UnitPrice = r.Next(1, 50)})
        End Function

        <Command>
        Public Sub ValidateCell(ByVal args As RowCellValidationArgs)
            If Not Equals(args.FieldName, NameOf(Product.UnitPrice)) OrElse Not CType(args.Item, Product).Discontinued Then Return
            Dim cellValue = CDbl(args.OldValue)
            Dim discount = 100 - Convert.ToDouble(args.NewValue) / cellValue * 100
            If discount > 0 AndAlso discount <= 30 Then Return
            Dim [error] = If(discount < 0, $"The price cannot be greater than {cellValue}", $"The discount cannot be greater than 30% ({cellValue * 0.7}). Please correct the price.")
            args.Result = New ValidationErrorInfo([error], ValidationErrorType.Critical)
        End Sub
    End Class
End Namespace
